pkg/apis/ceph.rook.io/v1/types.go

```diff
@@ -2320,6 +2320,7 @@ type NetworkSpec struct {
 	// Provider is what provides network connectivity to the cluster e.g. "host" or "multus".
 	// If the Provider is updated from being empty to "host" on a running cluster, then the operator will automatically fail over all the mons to apply the "host" network settings.
 	// +kubebuilder:validation:XValidation:message="network provider must be disabled (reverted to empty string) before a new provider is enabled",rule="self == '' || self == oldSelf"
+	// +kubebuilder:validation:XValidation:message="the legacy hostNetwork  setting can only be set if the network.provider is set to the empty  string",rule="(self.hostNetwork != true) || (self ==  '')"
```

Comment:
```suggestion
	// +kubebuilder:validation:XValidation:message="the legacy hostNetwork  setting can only be set if the network.provider is set to the empty  string",rule="(self.hostNetwork != true) || (self ==  '')"
```

The rule should be placed before L2319 and probably this`(self.hostNetwork != true) || (self == '')` need to change to `(self.hostNetwork == true) && (self.provider != '')` IMO

Reply:
Feel free to go thorugh https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules and https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/ for hints

Reply:
@subhamkrai I implemented your suggestion, but ci is still failing in multiple checks in different ways. 

And locally, I can't even get `cluster-test.yaml` created successfully:


here is the bottom of the output of `kubectl describe -n rook-ceph cephclusters my-cluster`: 
```
Events:
  Type     Reason           Age                   From                          Message
  ----     ------           ----                  ----                          -------
  Warning  ReconcileFailed  9m54s (x18 over 20m)  rook-ceph-cluster-controller  failed to reconcile CephCluster "rook-ceph/my-cluster". failed to add finalizer: failed to add finalizer "cephcluster.ceph.rook.io" on "my-cluster": CephCluster.ceph.rook.io "my-cluster" is invalid: spec.network: Invalid value: "object": no such key: hostNetwork evaluating rule: the legacy hostNetwork  setting can only be set if the network.provider is set to the empty string
```
I am surprised to see my newly  added error message here, because the (unmodified)  `cluster-test.yaml` does not even set `hostNetwork` or `provider` at all ...

Reply:
@subhamkrai  wrote:
> The rule should be placed before L2319 and probably this`(self.hostNetwork != true) || (self == '')` need to change to `(self.hostNetwork == true) && (self.provider != '')` IMO

Note that `&&` is the wrong logic here: it has to be `||`: We want to declare setting a nonemty provider invalid if `hostNetwork` enabled. Your suggested rule seems to enforce this invalid combination instead.

Reply:
I was thinking if both setting is provided through error

Reply:
@subhamkrai  wrote:
> I was thinking if both setting is provided through error

It is not clear to me what error you are expecting. The `cluster-test.yaml` was the unmodified version from the rook tree and it contains none of the two settings. So it could only have been an error in code? 🤷🏼